### ISSUE TITLE
A system of post-processing "side effects" for Hydra jobs

### ISSUE THREAD

**Issue by copumpkin** [NixOS/hydra#242]
An example of what I want: my Hydra instance produces images from Nix jobs; those images should be uploaded to EC2 and made into AMIs for potential deployment

The key consideration here is that Nix jobs are "pure", and run in a chroot/namespaced sandbox that can't talk to the network to upload the AMI. I don't want to change that, but it would be lovely to be able to  say "take the output of this Nix job and run the following script on it [ideally, in a slightly less restricted environment]"

Then we'd have a simpler "image -> AMI publisher script" that only needs to be able to talk to an AWS API endpoint, have some credentials, and otherwise needs no permissions to get at the disk or anywhere else on the network. Whenever the Hydra job runs, the effectful script would publish the AMI to Amazon and everyone would be happy.

A similar idea could move use closer to https://github.com/NixOS/hydra/issues/172, too.

**Comment by copumpkin** [comment]
Perhaps this would make more sense independently of Hydra, since it'd be ideal to be able to run these post-tasks on our own computers as well as on Hydra.

@edolstra @rbvermaa @shlevy any of you have thoughts on how to achieve this today, or about other good ways to make it happen?

It would be lovely if we could have a NixOS-unstable AMI that stays up to date with the channel, or even independently of channel updates on a per-job basis.

**Comment by copumpkin** [comment]
Perhaps a "post-success hook" concept might be the simplest first step towards this. It could simply send information about the successful build to an HTTP address of my choice, and then that can do what it wants with the result.

**Comment by edolstra** [comment]
Hydra's plugin mechanism already provides a `buildFinished` hook that could be used for this. For some use cases (like NixOS channel updating), an `evalFinished` hook might be more useful. And we could provide a generic plugin that sends HTTP POST requests on `buildFinished` / `evalFinished` (I vaguely remember that there is an issue for that).

**Comment by copumpkin** [comment]
Are those plugins documented somewhere? Seems very handy!

**Comment by edolstra** [comment]
No, but you can look at `EmailNotification.pm`, `S3Backup.pm` and others for examples.

Note: one problem with such plugins is that if they get interrupted/killed, then they won't be called again for that particular build (since the build is already registered as "finished" in the database). So you would lose an event.

**Comment by copumpkin** [comment]
Ah, how common are interruptions of that sort? My primary use case is to register AMIs on Amazon, but I'd probably also want to update a private channel and a few other things if that worked nicely.

**Comment by shlevy** [comment]
@edolstra Would it be very involved to put a 'buildFinished' queue into hydra (and add an idempotency expectation to plugins)?

**Comment by shlevy** [comment]
s/into hydra/into the db

**Comment by edolstra** [comment]
No, wouldn't be hard. We could just add a `notified` boolean to the Build table and have `hydra-queue-runner` process finished-but-not-yet-notified builds on startup.
